File: shared/BlacklistService.cs
using Dapper;
using Microsoft.Data.Sqlite;

namespace shared.Services;

public class BlacklistService
{
    private readonly string _connectionString;

    public BlacklistService(FileService fileService)
    {
        _connectionString = fileService.ConnectionString;
        InitDb();
    }

    private void InitDb()
    {
        using var conn = new SqliteConnection(_connectionString);
        conn.Execute("""
                         CREATE TABLE IF NOT EXISTS Blacklist (
                             Ip        TEXT PRIMARY KEY,
                             AddedAt   TEXT NOT NULL
                         )
                     """);
    }

    public async Task<bool> IsBlockedAsync(string ip)
    {
        using var conn = new SqliteConnection(_connectionString);
        var count = await conn.ExecuteScalarAsync<int>(
            "SELECT COUNT(*) FROM Blacklist WHERE Ip = @Ip", new { Ip = ip });
        return count > 0;
    }

    public async Task AddAsync(string ip)
    {
        using var conn = new SqliteConnection(_connectionString);
        await conn.ExecuteAsync("""
                                    INSERT OR IGNORE INTO Blacklist (Ip, AddedAt)
                                    VALUES (@Ip, @AddedAt)
                                """, new { Ip = ip, AddedAt = DateTime.UtcNow.ToString("O") });
    }

    public async Task RemoveAsync(string ip)
    {
        using var conn = new SqliteConnection(_connectionString);
        await conn.ExecuteAsync("DELETE FROM Blacklist WHERE Ip = @Ip", new { Ip = ip });
    }

    public async Task<IEnumerable<string>> ListAsync()
    {
        using var conn = new SqliteConnection(_connectionString);
        return await conn.QueryAsync<string>("SELECT Ip FROM Blacklist ORDER BY AddedAt DESC");
    }
}

File: src/main/java/dev/uncognic/shared/service/BlacklistService.java
package dev.uncognic.shared.service;

import jakarta.annotation.PostConstruct;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.List;

@Service
public class BlacklistService {
    private final JdbcTemplate db;

    public BlacklistService(JdbcTemplate db) {
        this.db = db;
    }

    @PostConstruct
    public void initDb() {
        db.execute("""
            CREATE TABLE IF NOT EXISTS Blacklist (
                Ip TEXT PRIMARY KEY,
                AddedAt TEXT NOT NULL
            )
            """);
    }

    // check if an ip is in the table
    public boolean isBlocked(String ip) {
        Integer count = db.queryForObject("SELECT COUNT(*) FROM Blacklist WHERE Ip = ?", Integer.class, ip);
        return count != null && count > 0;
    }

    // add into the table
    public void add(String ip) {
        db.update("INSERT OR IGNORE INTO Blacklist (Ip, AddedAt) VALUES (?, ?)",
            ip, Instant.now().toString());
    }

    // remove
    public void remove(String ip) {
        db.update("DELETE FROM Blacklist WHERE Ip = ?", ip);
    }

    // list all
    public List<String> list() {
        return db.queryForList("SELECT Ip FROM Blacklist ORDER BY AddedAt DESC", String.class);
    }
}
